File: ios/BB8Kit/Sources/BB8Kit/Sensors.swift
import Foundation

/// Sensor stream masks and sample decoding.
///
/// BB-8 streams samples as a flat sequence of big-endian **signed 16-bit**
/// values. There are no field tags on the wire: the receiver must know which
/// sensors it requested and reconstruct the layout, in **descending mask-bit
/// order**, primary-mask fields first and extended-mask fields after.
///
/// That implicit contract is why this type exists. Get the ordering wrong and
/// every field decodes as its neighbour — plausible-looking numbers that are
/// entirely wrong, which is far worse than a crash.
public enum SpheroSensors {

    public struct Spec: Sendable, Hashable {
        public let name: String
        public let bit: UInt32
        public let scale: Double
        public let unit: String

        init(_ name: String, _ bit: UInt32, scale: Double = 1, unit: String = "") {
            self.name = name; self.bit = bit; self.scale = scale; self.unit = unit
        }
    }

    /// Fields selected by the primary 32-bit mask, descending bit order.
    public static let primary: [Spec] = [
        Spec("pitch", 0x0004_0000, unit: "deg"),
        Spec("roll", 0x0002_0000, unit: "deg"),
        Spec("yaw", 0x0001_0000, unit: "deg"),
        Spec("accelX", 0x0000_8000, scale: 1.0 / 4096, unit: "g"),
        Spec("accelY", 0x0000_4000, scale: 1.0 / 4096, unit: "g"),
        Spec("accelZ", 0x0000_2000, scale: 1.0 / 4096, unit: "g"),
        Spec("gyroX", 0x0000_1000, scale: 0.1, unit: "deg/s"),
        Spec("gyroY", 0x0000_0800, scale: 0.1, unit: "deg/s"),
        Spec("gyroZ", 0x0000_0400, scale: 0.1, unit: "deg/s"),
        Spec("emfLeft", 0x0000_0040),
        Spec("emfRight", 0x0000_0020),
    ]

    /// Fields selected by the extended 32-bit mask, descending bit order.
    public static let extended: [Spec] = [
        Spec("quatX", 0x8000_0000, scale: 1.0 / 10000),
        Spec("quatY", 0x4000_0000, scale: 1.0 / 10000),
        Spec("quatZ", 0x2000_0000, scale: 1.0 / 10000),
        Spec("quatW", 0x1000_0000, scale: 1.0 / 10000),
        Spec("locatorX", 0x0800_0000, unit: "cm"),
        Spec("locatorY", 0x0400_0000, unit: "cm"),
        Spec("accelOne", 0x0200_0000),
        Spec("velocityX", 0x0100_0000, scale: 0.1, unit: "cm/s"),
        Spec("velocityY", 0x0080_0000, scale: 0.1, unit: "cm/s"),
        Spec("speed", 0x0040_0000, unit: "cm/s"),
    ]

    /// Enough to see motion and position without saturating the link.
    public static let drivePreset = ["yaw", "speed", "locatorX", "locatorY", "velocityX", "velocityY"]

    /// The firmware's sensor sample clock.
    ///
    /// The streaming command takes a **divisor** of this, not a period in
    /// milliseconds — verified on hardware to within 0.25% (divisor 100 →
    /// 4.01 Hz, 50 → 8.01 Hz, 25 → 15.96 Hz). Reading it as milliseconds
    /// understates the rate by 25x and makes a healthy stream look like heavy
    /// packet loss.
    public static let baseHz: Double = 400

    public static func divisor(forHz hz: Double) -> Int {
        max(1, min(0xFFFF, Int((baseHz / hz).rounded())))
    }

    public static func hz(forDivisor divisor: Int) -> Double {
        divisor == 0 ? 0 : baseHz / Double(divisor)
    }

    public enum SensorError: Error, Equatable {
        case unknownSensor(String)
        case badLength(expected: Int, got: Int)
    }

    /// Turn sensor names into the `(primary, extended)` mask pair.
    ///
    /// Throws on an unrecognised name rather than skipping it: a dropped field
    /// would shift every subsequent value in the decoded sample.
    public static func masks(for names: [String]) throws -> (primary: UInt32, extended: UInt32) {
        var p: UInt32 = 0, e: UInt32 = 0
        for name in names {
            if let spec = primary.first(where: { $0.name == name }) {
                p |= spec.bit
            } else if let spec = extended.first(where: { $0.name == name }) {
                e |= spec.bit
            } else {
                throw SensorError.unknownSensor(name)
            }
        }
        return (p, e)
    }

    /// The field order a sample with these masks will arrive in.
    public static func layout(primary p: UInt32, extended e: UInt32) -> [Spec] {
        primary.filter { $0.bit & p != 0 } + extended.filter { $0.bit & e != 0 }
    }

    /// Decode one scaled sample.
    ///
    /// The length check is the only defence against a mask/payload mismatch:
    /// with no field tags on the wire, a wrong mask yields confidently wrong
    /// numbers rather than an error. Fail loudly instead.
    public static func decodeSample(
        _ payload: [UInt8], primary p: UInt32, extended e: UInt32
    ) throws -> [String: Double] {
        let fields = layout(primary: p, extended: e)
        guard payload.count == fields.count * 2 else {
            throw SensorError.badLength(expected: fields.count * 2, got: payload.count)
        }
        var out: [String: Double] = [:]
        for (index, spec) in fields.enumerated() {
            let raw = Int16(bitPattern: UInt16(payload[index * 2]) << 8 | UInt16(payload[index * 2 + 1]))
            out[spec.name] = Double(raw) * spec.scale
        }
        return out
    }

    /// Decode an async `0x03` payload, which may carry several samples.
    public static func decodeFrame(
        _ payload: [UInt8], primary p: UInt32, extended e: UInt32
    ) throws -> [[String: Double]] {
        let size = layout(primary: p, extended: e).count * 2
        guard size > 0 else { return [] }
        guard payload.count % size == 0 else {
            throw SensorError.badLength(expected: size, got: payload.count)
        }
        return try stride(from: 0, to: payload.count, by: size).map {
            try decodeSample(Array(payload[$0 ..< $0 + size]), primary: p, extended: e)
        }
    }
}

extension Sphero {
    /// Start (or stop, with `mask: 0`) async sensor streaming.
    ///
    /// - Parameter divisor: divides `SpheroSensors.baseHz`. Rate is
    ///   `400 / divisor` Hz. Use `SpheroSensors.divisor(forHz:)` rather than
    ///   passing a period in milliseconds.
    public static func setDataStreaming(
        divisor: Int, samplesPerPacket: Int = 1, mask: UInt32,
        count: UInt8 = 0, extendedMask: UInt32 = 0, seq: UInt8 = 0
    ) -> Data {
        var data: [UInt8] = []
        data += [UInt8((divisor >> 8) & 0xFF), UInt8(divisor & 0xFF)]
        data += [UInt8((samplesPerPacket >> 8) & 0xFF), UInt8(samplesPerPacket & 0xFF)]
        data += [UInt8((mask >> 24) & 0xFF), UInt8((mask >> 16) & 0xFF),
                 UInt8((mask >> 8) & 0xFF), UInt8(mask & 0xFF)]
        data += [count]
        data += [UInt8((extendedMask >> 24) & 0xFF), UInt8((extendedMask >> 16) & 0xFF),
                 UInt8((extendedMask >> 8) & 0xFF), UInt8(extendedMask & 0xFF)]
        return build(did: .sphero, cid: SpheroCommand.setDataStreaming.rawValue,
                     seq: seq, data: data, answer: true)
    }
}

File: src/bb8ctl/sensors.py
"""Sensor stream masks and sample decoding for BB-8.

Pure, like :mod:`bb8ctl.protocol` -- no I/O, no platform dependency, so it ports
to Swift alongside the codec (requirement N8).

BB-8 streams samples as a flat sequence of big-endian **signed 16-bit** values.
There are no field tags on the wire: the receiver must know which sensors were
requested and reconstruct the layout. Values appear in **descending mask-bit
order**, all primary-mask fields first, then all extended-mask fields.

That implicit contract is the whole reason this module exists. Get the ordering
wrong and every field silently decodes as its neighbour -- plausible-looking
numbers that are entirely wrong, which is far worse than a crash.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Callable, Iterable


@dataclass(frozen=True)
class SensorSpec:
    """One streamable field."""

    name: str
    bit: int
    #: Converts the raw int16 to a physical unit.
    scale: Callable[[int], float] = float
    unit: str = ""


def _div(n: float) -> Callable[[int], float]:
    return lambda v: v / n


def _mul(n: float) -> Callable[[int], float]:
    return lambda v: v * n


#: Fields selected by the primary 32-bit mask, descending bit order.
PRIMARY: tuple[SensorSpec, ...] = (
    SensorSpec("accel_raw_x", 0x00400000, unit="raw"),
    SensorSpec("accel_raw_y", 0x00200000, unit="raw"),
    SensorSpec("accel_raw_z", 0x00100000, unit="raw"),
    SensorSpec("pitch", 0x00040000, unit="deg"),
    SensorSpec("roll", 0x00020000, unit="deg"),
    SensorSpec("yaw", 0x00010000, unit="deg"),
    SensorSpec("accel_x", 0x00008000, _div(4096), "g"),
    SensorSpec("accel_y", 0x00004000, _div(4096), "g"),
    SensorSpec("accel_z", 0x00002000, _div(4096), "g"),
    SensorSpec("gyro_x", 0x00001000, _mul(0.1), "deg/s"),
    SensorSpec("gyro_y", 0x00000800, _mul(0.1), "deg/s"),
    SensorSpec("gyro_z", 0x00000400, _mul(0.1), "deg/s"),
    SensorSpec("emf_left", 0x00000040, unit="raw"),
    SensorSpec("emf_right", 0x00000020, unit="raw"),
)

#: Fields selected by the extended 32-bit mask, descending bit order.
EXTENDED: tuple[SensorSpec, ...] = (
    SensorSpec("quat_x", 0x80000000, _div(10000)),
    SensorSpec("quat_y", 0x40000000, _div(10000)),
    SensorSpec("quat_z", 0x20000000, _div(10000)),
    SensorSpec("quat_w", 0x10000000, _div(10000)),
    SensorSpec("locator_x", 0x08000000, unit="cm"),
    SensorSpec("locator_y", 0x04000000, unit="cm"),
    SensorSpec("accel_one", 0x02000000, unit="raw"),
    SensorSpec("velocity_x", 0x01000000, _mul(0.1), "cm/s"),
    SensorSpec("velocity_y", 0x00800000, _mul(0.1), "cm/s"),
    SensorSpec("speed", 0x00400000, unit="cm/s"),
)

BY_NAME: dict[str, SensorSpec] = {s.name: s for s in (*PRIMARY, *EXTENDED)}

#: A useful default for driving: enough to see motion and position without
#: saturating the link. Locator and speed are what the trajectory work needs.
DRIVE_PRESET: tuple[str, ...] = (
    "yaw", "speed", "locator_x", "locator_y", "velocity_x", "velocity_y",
)


class UnknownSensor(KeyError):
    """Raised for a sensor name not in :data:`BY_NAME`."""


def build_masks(names: Iterable[str]) -> tuple[int, int]:
    """Turn sensor names into the ``(primary, extended)`` mask pair.

    Raises :class:`UnknownSensor` rather than silently skipping a typo -- a
    dropped field would shift every subsequent value in the decoded sample.
    """
    primary = extended = 0
    primary_bits = {s.name: s.bit for s in PRIMARY}
    for name in names:
        if name in primary_bits:
            primary |= primary_bits[name]
        elif name in {s.name for s in EXTENDED}:
            extended |= BY_NAME[name].bit
        else:
            raise UnknownSensor(f"unknown sensor {name!r}")
    return primary, extended


def layout(primary: int, extended: int) -> tuple[SensorSpec, ...]:
    """The field order a sample with these masks will arrive in."""
    return tuple(s for s in PRIMARY if s.bit & primary) + tuple(
        s for s in EXTENDED if s.bit & extended
    )


def expected_length(primary: int, extended: int) -> int:
    """Bytes in one sample. Two per field, since every value is an int16."""
    return 2 * len(layout(primary, extended))


class SampleLengthError(ValueError):
    """Payload length disagrees with the requested masks."""


def decode_sample(data: bytes, primary: int, extended: int) -> dict[str, float]:
    """Decode one scaled sample.

    The length check is the only defence against a mask/payload mismatch: with
    no field tags on the wire, a wrong mask produces confidently wrong numbers
    rather than an error. Fail loudly instead.
    """
    fields = layout(primary, extended)
    if len(data) != 2 * len(fields):
        raise SampleLengthError(
            f"expected {2 * len(fields)} bytes for {len(fields)} fields, got {len(data)}"
        )
    out: dict[str, float] = {}
    for i, spec in enumerate(fields):
        raw = int.from_bytes(data[i * 2 : i * 2 + 2], "big", signed=True)
        out[spec.name] = spec.scale(raw)
    return out


def decode_frame(data: bytes, primary: int, extended: int) -> list[dict[str, float]]:
    """Decode an async 0x03 payload, which may carry several samples."""
    size = expected_length(primary, extended)
    if size == 0:
        return []
    if len(data) % size:
        raise SampleLengthError(f"payload {len(data)} is not a multiple of sample size {size}")
    return [decode_sample(data[i : i + size], primary, extended) for i in range(0, len(data), size)]
